<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro">

	<xacro:include filename="inertial_macros.xacro"/>

	<xacro:property name="damping" value="10"/>
	<xacro:property name="friction" value="500"/>	
	<xacro:property name="mu1" value="550"/>
	<xacro:property name="mu2" value="550"/>
	<xacro:property name="velocity" value="1"/>
	<xacro:property name="effort" value="10"/>
	<xacro:property name="base_mass" value="10"/>
	<xacro:property name="leg_mass" value="5"/>
	<xacro:property name="leg_mass3" value="5"/>
	<xacro:property name="fdir" value="1 0 0"/>

	<link name="base_link">

	</link>
	
	<!-- STAND -->
	<joint name="stand_joint" type="fixed">
		<parent link="base_link"/>
		<child link="stand_link"/>
		<origin xyz="0 0 0"/>
	</joint>
	
	<link name="stand_link">
		<visual>
			<origin xyz="0 0 0.02"/>
			<geometry>
				<box size="0.5 0.5 0.04"/>
			</geometry>
		</visual>			

		<collision>
			<origin xyz="0 0 0.02"/>
			<geometry>
				<box size="0.5 0.5 0.04"/>
			</geometry>
		</collision>

		<xacro:inertial_box mass="${base_mass}" x="0.5" y="0.5" z="0.04">
			<origin xyz="0 0 0.02" rpy="0 0 0"/>
		</xacro:inertial_box>
	</link>
	
	<gazebo reference="stand_link">
		<material>Gazebo/Red</material>
		<mu1 value="${mu1}"/>
		<mu2 value="${mu2}"/>
		<fdir1 value="${fdir}"/>
		<maxVel>1.0</maxVel>
    	<minDepth>0.01</minDepth>
	</gazebo>
	
	<!-- LEFT_LEG1_AX1 -->
	<joint name="lleg1_ax1_joint" type="revolute">
		<parent link="stand_link"/>
		<child link="lleg1_ax1_link"/>
		<origin xyz="0.22 0.22 0.12"/>
		<limit effort="${effort}" velocity="${velocity}" lower="${-pi/2}" upper="${pi/2}"/>
		<axis xyz="0 0 1"/>
		<dynamics damping="${damping}" friction="${friction}"/>
	</joint>
	
	<link name="lleg1_ax1_link">
		<visual>
			<origin xyz="0 0 -0.03"/>
			<geometry>
				<box size="0.04 0.04 0.1"/>
			</geometry>
		</visual>

		<collision>
			<origin xyz="0 0 -0.03"/>
			<geometry>
				<box size="0.04 0.04 0.1"/>
			</geometry>
		</collision>
		
		<xacro:inertial_box mass="${leg_mass}" x="0.04" y="0.04" z="0.1">
			<origin xyz="0 0 -0.03" rpy="0 0 0"/>
		</xacro:inertial_box>
	</link>

	<gazebo reference="lleg1_ax1_link">
		<material>Gazebo/Black</material>
		<mu1 value="${mu1}"/>
		<mu2 value="${mu2}"/>
		<fdir1 value="${fdir}"/>
		<maxVel>1.0</maxVel>
    	<minDepth>0.01</minDepth>
	</gazebo>

	<!-- LEFT_LEG1_AX2 -->
	<joint name="lleg1_ax2_joint" type="revolute">
		<parent link="lleg1_ax1_link"/>
		<child link="lleg1_ax2_link"/>
		<origin xyz="0 0.04 0" rpy="0 ${pi/2} 0"/>
		<limit effort="${effort}" velocity="${velocity}" lower="${-pi/2}" upper="${pi/2}"/>
		<axis xyz="0 1 0"/>
		<dynamics damping="${damping}" friction="${friction}"/>
	</joint>
	
	<link name="lleg1_ax2_link">
		<visual>
			<origin xyz="0 0 0.18"/>
			<geometry>
				<box size="0.04 0.04 0.4"/>
			</geometry>
		</visual>

		<collision>
			<origin xyz="0 0 0.18"/>
			<geometry>
				<box size="0.04 0.04 0.4"/>
			</geometry>
		</collision>

		<xacro:inertial_box mass="${leg_mass}" x="0.04" y="0.04" z="0.4">
			<origin xyz="0 0 0.18" rpy="0 0 0"/>
		</xacro:inertial_box>
	</link>

	<gazebo reference="lleg1_ax2_link">
		<material>Gazebo/Red</material>
		<mu1 value="${mu1}"/>
		<mu2 value="${mu2}"/>
		<fdir1 value="${fdir}"/>
		<maxVel>1.0</maxVel>
    	<minDepth>0.01</minDepth>
	</gazebo>

	<!-- LEFT_LEG1_AX3 -->
	<joint name="lleg1_ax3_joint" type="revolute">
		<parent link="lleg1_ax2_link"/>
		<child link="lleg1_ax3_link"/>
		<origin xyz="0 0 0.36"/>
		<limit effort="${effort}" velocity="${velocity}" lower="${-pi}" upper="${pi}"/>
		<axis xyz="0 1 0"/>
		<dynamics damping="${damping}" friction="${friction}"/>
	</joint>
	
	<link name="lleg1_ax3_link">
		<visual>
			<origin xyz="0 -0.04 0.18"/>
			<geometry>
				<box size="0.04 0.04 0.4"/>
			</geometry>
		</visual>

		<collision>
			<origin xyz="0 -0.04 0.18"/>
			<geometry>
				<box size="0.04 0.04 0.4"/>
			</geometry>
		</collision>

		<xacro:inertial_box mass="${leg_mass3}" x="0.04" y="0.04" z="0.4">
			<origin xyz="0 -0.04 0.18" rpy="0 0 0"/>
		</xacro:inertial_box>
	</link>

	<gazebo reference="lleg1_ax3_link">
		<material>Gazebo/Black</material>
		<mu1 value="${mu1}"/>
		<mu2 value="${mu2}"/>
		<fdir1 value="${fdir}"/>
		<maxVel>1.0</maxVel>
    	<minDepth>0.01</minDepth>
		<kp>1000000.0</kp>
		<kd>100.0</kd>
	</gazebo>

	<!-- LEFT_LEG2_AX1 -->
	<joint name="lleg2_ax1_joint" type="revolute">
		<parent link="stand_link"/>
		<child link="lleg2_ax1_link"/>
		<origin xyz="0.22 -0.22 0.12"/>
		<limit effort="${effort}" velocity="${velocity}" lower="${-pi/2}" upper="${pi/2}"/>
		<axis xyz="0 0 -1"/>
		<dynamics damping="${damping}" friction="${friction}"/>
	</joint>
	
	<link name="lleg2_ax1_link">
		<visual>
			<origin xyz="0 0 -0.03"/>
			<geometry>
				<box size="0.04 0.04 0.1"/>
			</geometry>
		</visual>

		<collision>
			<origin xyz="0 0 -0.03"/>
			<geometry>
				<box size="0.04 0.04 0.1"/>
			</geometry>
		</collision>

		<xacro:inertial_box mass="${leg_mass}" x="0.04" y="0.04" z="0.1">
			<origin xyz="0 0 -0.03" rpy="0 0 0"/>
		</xacro:inertial_box>
	</link>

	<gazebo reference="lleg2_ax1_link">
		<material>Gazebo/Black</material>
		<mu1 value="${mu1}"/>
		<mu2 value="${mu2}"/>
		<fdir1 value="${fdir}"/>
		<maxVel>1.0</maxVel>
    	<minDepth>0.01</minDepth>
	</gazebo>
	
	<!-- LEFT_LEG2_AX2 -->
	<joint name="lleg2_ax2_joint" type="revolute">
		<parent link="lleg2_ax1_link"/>
		<child link="lleg2_ax2_link"/>
		<origin xyz="0 0.04 0" rpy="0 ${pi/2} 0"/>
		<limit effort="${effort}" velocity="${velocity}" lower="${-pi/2}" upper="${pi/2}"/>
		<axis xyz="0 1 0"/>
		<dynamics damping="${damping}" friction="${friction}"/>
	</joint>
	
	<link name="lleg2_ax2_link">
		<visual>
			<origin xyz="0 0 0.18"/>
			<geometry>
				<box size="0.04 0.04 0.4"/>
			</geometry>
		</visual>

		<collision>
			<origin xyz="0 0 0.18"/>
			<geometry>
				<box size="0.04 0.04 0.4"/>
			</geometry>
		</collision>
		
		<xacro:inertial_box mass="${leg_mass}" x="0.04" y="0.04" z="0.4">
			<origin xyz="0 0 0.18" rpy="0 0 0"/>
		</xacro:inertial_box>
	</link>

	<gazebo reference="lleg2_ax2_link">
		<material>Gazebo/Red</material>
		<mu1 value="${mu1}"/>
		<mu2 value="${mu2}"/>
		<fdir1 value="${fdir}"/>
		<maxVel>1.0</maxVel>
    	<minDepth>0.01</minDepth>
	</gazebo>

	<!-- LEFT_LEG2_AX3 -->
	<joint name="lleg2_ax3_joint" type="revolute">
		<parent link="lleg2_ax2_link"/>
		<child link="lleg2_ax3_link"/>
		<origin xyz="0 0 0.36"/>
		<limit effort="${effort}" velocity="${velocity}" lower="${-pi}" upper="${pi}"/>
		<axis xyz="0 1 0"/>
		<dynamics damping="${damping}" friction="${friction}"/>
	</joint>
	
	<link name="lleg2_ax3_link">
		<visual>
			<origin xyz="0 -0.04 0.18"/>
			<geometry>
				<box size="0.04 0.04 0.4"/>
			</geometry>
		</visual>

		<collision>
			<origin xyz="0 -0.04 0.18"/>
			<geometry>
				<box size="0.04 0.04 0.4"/>
			</geometry>
		</collision>

		<!-- added pads 
		<visual>
			<origin xyz="0 -0.04 0.38"/>
			<geometry>
				<sphere radius="0.04"/>
			</geometry>
		</visual>

		<collision>
			<origin xyz="0 -0.04 0.38"/>
			<geometry>
				<sphere radius="0.04"/>
			</geometry>
		</collision>	
		-->
		<xacro:inertial_box mass="${leg_mass3}" x="0.04" y="0.04" z="0.4">
			<origin xyz="0 -0.04 0.18" rpy="0 0 0"/>
		</xacro:inertial_box>
	</link>

	<gazebo reference="lleg2_ax3_link">
		<material>Gazebo/Black</material>
		<mu1 value="${mu1}"/>
		<mu2 value="${mu2}"/>
		<fdir1 value="${fdir}"/>
		<maxVel>1.0</maxVel>
    	<minDepth>0.01</minDepth>
		<kp>1000000.0</kp>
		<kd>100.0</kd>
	</gazebo>
	
	<!-- RIGHT_LEG1_AX1 -->
	<joint name="rleg1_ax1_joint" type="revolute">
		<parent link="stand_link"/>
		<child link="rleg1_ax1_link"/>
		<origin xyz="-0.22 0.22 0.12"/>
		<limit effort="${effort}" velocity="${velocity}" lower="${-pi/2}" upper="${pi/2}"/>
		<axis xyz="0 0 -1"/>
		<dynamics damping="${damping}" friction="${friction}"/>
	</joint>
	
	<link name="rleg1_ax1_link">
		<visual>
			<origin xyz="0 0 -0.03"/>
			<geometry>
				<box size="0.04 0.04 0.1"/>
			</geometry>
		</visual>
		
		<collision>
			<origin xyz="0 0 -0.03"/>
			<geometry>
				<box size="0.04 0.04 0.1"/>
			</geometry>
		</collision>
		
		<xacro:inertial_box mass="${leg_mass}" x="0.04" y="0.04" z="0.1">
			<origin xyz="0 0 -0.03" rpy="0 0 0"/>
		</xacro:inertial_box>
	</link>

	<gazebo reference="rleg1_ax1_link">
		<material>Gazebo/Black</material>
		<mu1 value="${mu1}"/>
		<mu2 value="${mu2}"/>
		<fdir1 value="${fdir}"/>
		<maxVel>1.0</maxVel>
    	<minDepth>0.01</minDepth>
	</gazebo>	

	<!-- RIGHT_LEG1_AX2 -->
	<joint name="rleg1_ax2_joint" type="revolute">
		<parent link="rleg1_ax1_link"/>
		<child link="rleg1_ax2_link"/>
		<origin xyz="0 0.04 0" rpy="0 ${-pi/2} 0"/>
		<limit effort="${effort}" velocity="${velocity}" lower="${-pi/2}" upper="${pi/2}"/>
		<axis xyz="0 -1 0"/>
		<dynamics damping="${damping}" friction="${friction}"/>
	</joint>
	
	<link name="rleg1_ax2_link">
		<visual>
			<origin xyz="0 0 0.18"/>
			<geometry>
				<box size="0.04 0.04 0.4"/>
			</geometry>
		</visual>

		<collision>
			<origin xyz="0 0 0.18"/>
			<geometry>
				<box size="0.04 0.04 0.4"/>
			</geometry>
		</collision>

		<xacro:inertial_box mass="${leg_mass}" x="0.04" y="0.04" z="0.4">
			<origin xyz="0 0 0.18" rpy="0 0 0"/>
		</xacro:inertial_box>
	</link>

	<gazebo reference="rleg1_ax2_link">
		<material>Gazebo/Red</material>
		<mu1 value="${mu1}"/>
		<mu2 value="${mu2}"/>
		<fdir1 value="${fdir}"/>
		<maxVel>1.0</maxVel>
    	<minDepth>0.01</minDepth>
	</gazebo>

	<!-- RIGHT_LEG1_AX3 -->
	<joint name="rleg1_ax3_joint" type="revolute">
		<parent link="rleg1_ax2_link"/>
		<child link="rleg1_ax3_link"/>
		<origin xyz="0 0 0.36"/>
		<limit effort="${effort}" velocity="${velocity}" lower="${-pi}" upper="${pi}"/>
		<axis xyz="0 -1 0"/>
		<dynamics damping="${damping}" friction="${friction}"/>
	</joint>
	
	<link name="rleg1_ax3_link">
		<visual>
			<origin xyz="0 -0.04 0.18"/>
			<geometry>
				<box size="0.04 0.04 0.4"/>
			</geometry>
		</visual>

		<collision>
			<origin xyz="0 -0.04 0.18"/>
			<geometry>
				<box size="0.04 0.04 0.4"/>
			</geometry>
		</collision>
		
		<!-- added pads 
		<visual>
			<origin xyz="0 -0.04 0.38"/>
			<geometry>
				<sphere radius="0.04"/>
			</geometry>
		</visual>

		<collision>
			<origin xyz="0 -0.04 0.38"/>
			<geometry>
				<sphere radius="0.04"/>
			</geometry>
		</collision>
		-->
		<xacro:inertial_box mass="${leg_mass3}" x="0.04" y="0.04" z="0.4">
			<origin xyz="0 -0.04 0.18" rpy="0 0 0"/>
		</xacro:inertial_box>
	</link>

	<gazebo reference="rleg1_ax3_link">
		<material>Gazebo/Black</material>
		<mu1 value="${mu1}"/>
		<mu2 value="${mu2}"/>
		<fdir1 value="${fdir}"/>
		<maxVel>1.0</maxVel>
    	<minDepth>0.01</minDepth>
		<kp>1000000.0</kp>
		<kd>100.0</kd>
	</gazebo>

	<!-- RIGHT_LEG2_AX1 -->
	<joint name="rleg2_ax1_joint" type="revolute">
		<parent link="stand_link"/>
		<child link="rleg2_ax1_link"/>
		<origin xyz="-0.22 -0.22 0.12"/>
		<limit effort="${effort}" velocity="${velocity}" lower="${-pi/2}" upper="${pi/2}"/>
		<axis xyz="0 0 1"/>
		<dynamics damping="${damping}" friction="${friction}"/>
	</joint>
	
	<link name="rleg2_ax1_link">
		<visual>
			<origin xyz="0 0 -0.03"/>
			<geometry>
				<box size="0.04 0.04 0.1"/>
			</geometry>
		</visual>

		<collision>
			<origin xyz="0 0 -0.03"/>
			<geometry>
				<box size="0.04 0.04 0.1"/>
			</geometry>
		</collision>

		<xacro:inertial_box mass="${leg_mass}" x="0.04" y="0.04" z="0.1">
			<origin xyz="0 0 -0.03" rpy="0 0 0"/>
		</xacro:inertial_box>
	</link>
	
	<gazebo reference="rleg2_ax1_link">
		<material>Gazebo/Black</material>
		<mu1 value="${mu1}"/>
		<mu2 value="${mu2}"/>
		<fdir1 value="${fdir}"/>
		<maxVel>1.0</maxVel>
    	<minDepth>0.01</minDepth>
	</gazebo>

	<!-- RIGHT_LEG2_AX2 -->
	<joint name="rleg2_ax2_joint" type="revolute">
		<parent link="rleg2_ax1_link"/>
		<child link="rleg2_ax2_link"/>
		<origin xyz="0 0.04 0" rpy="0 ${-pi/2} 0"/>
		<limit effort="${effort}" velocity="${velocity}" lower="${-pi/2}" upper="${pi/2}"/>
		<axis xyz="0 -1 0"/>
		<dynamics damping="${damping}" friction="${friction}"/>
	</joint>
	
	<link name="rleg2_ax2_link">
		<visual>
			<origin xyz="0 0 0.18"/>
			<geometry>
				<box size="0.04 0.04 0.4"/>
			</geometry>
		</visual>

		<collision>
			<origin xyz="0 0 0.18"/>
			<geometry>
				<box size="0.04 0.04 0.4"/>
			</geometry>
		</collision>

		<xacro:inertial_box mass="${leg_mass}" x="0.04" y="0.04" z="0.4">
			<origin xyz="0 0 0.18" rpy="0 0 0"/>
		</xacro:inertial_box>
	</link>

	<gazebo reference="rleg2_ax2_link">
		<material>Gazebo/Red</material>
		<mu1 value="${mu1}"/>
		<mu2 value="${mu2}"/>
		<fdir1 value="${fdir}"/>
		<maxVel>1.0</maxVel>
    	<minDepth>0.01</minDepth>
	</gazebo>

	<!-- RIGHT_LEG2_AX3 -->
	<joint name="rleg2_ax3_joint" type="revolute">
		<parent link="rleg2_ax2_link"/>
		<child link="rleg2_ax3_link"/>
		<origin xyz="0 0 0.36"/>
		<limit effort="${effort}" velocity="${velocity}" lower="${-pi}" upper="${pi}"/>
		<axis xyz="0 -1 0"/>
		<dynamics damping="${damping}" friction="${friction}"/>
	</joint>
	
	<link name="rleg2_ax3_link">
		<visual>
			<origin xyz="0 -0.04 0.18"/>
			<geometry>
				<box size="0.04 0.04 0.4"/>
			</geometry>
		</visual>

		<collision>
			<origin xyz="0 -0.04 0.18"/>
			<geometry>
				<box size="0.04 0.04 0.4"/>
			</geometry>
		</collision>

		<!-- added pads 
		<visual>
			<origin xyz="0 -0.04 0.38"/>
			<geometry>
				<sphere radius="0.04"/>
			</geometry>
		</visual>

		<collision>
			<origin xyz="0 -0.04 0.38"/>
			<geometry>
				<sphere radius="0.04"/>
			</geometry>
		</collision>
		-->
		<xacro:inertial_box mass="${leg_mass3}" x="0.04" y="0.04" z="0.4">
			<origin xyz="0 -0.04 0.18" rpy="0 0 0"/>
		</xacro:inertial_box>
	</link>
	
	<gazebo reference="rleg2_ax3_link">
		<material>Gazebo/Black</material>
		<mu1 value="${mu1}"/>
		<mu2 value="${mu2}"/>
		<fdir1 value="${fdir}"/>
		<maxVel>1.0</maxVel>
    	<minDepth>0.01</minDepth>
		<kp>1000000.0</kp>
 		<kd>100.0</kd>
	</gazebo>

	<ros2_control name="GazeboSystem" type="system">
        <hardware>
            <plugin>gazebo_ros2_control/GazeboSystem</plugin>
        </hardware>

        <joint name="lleg1_ax1_joint">
            <command_interface name="position">
                <param name="min">${-pi/2}</param>
                <param name="max">${pi/2}</param>
            </command_interface>
            <state_interface name="position">
                <param name="initial_value">${0}</param>
            </state_interface>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>

		<joint name="lleg1_ax2_joint">
            <command_interface name="position">
                <param name="min">${-pi/2}</param>
                <param name="max">${pi/2}</param>
            </command_interface>
            <state_interface name="position">
                <param name="initial_value">${0}</param>
            </state_interface>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>

		<joint name="lleg1_ax3_joint">
            <command_interface name="position">
                <param name="min">${-pi}</param>
                <param name="max">${pi}</param>
            </command_interface>
            <state_interface name="position">
                <param name="initial_value">${0}</param>
            </state_interface>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>

		<joint name="lleg2_ax1_joint">
            <command_interface name="position">
                <param name="min">${-pi/2}</param>
                <param name="max">${pi/2}</param>
            </command_interface>
            <state_interface name="position">
                <param name="initial_value">${0}</param>
            </state_interface>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>
		
		<joint name="lleg2_ax2_joint">
            <command_interface name="position">
                <param name="min">${-pi/2}</param>
                <param name="max">${pi/2}</param>
            </command_interface>
            <state_interface name="position">
                <param name="initial_value">${0}</param>
            </state_interface>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>

		<joint name="lleg2_ax3_joint">
            <command_interface name="position">
                <param name="min">${-pi}</param>
                <param name="max">${pi}</param>
            </command_interface>
            <state_interface name="position">
                <param name="initial_value">${0}</param>
            </state_interface>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>

		<joint name="rleg1_ax1_joint">
            <command_interface name="position">
                <param name="min">${-pi/2}</param>
                <param name="max">${pi/2}</param>
            </command_interface>
            <state_interface name="position">
                <param name="initial_value">${0}</param>
            </state_interface>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>

		<joint name="rleg1_ax2_joint">
            <command_interface name="position">
                <param name="min">${-pi/2}</param>
                <param name="max">${pi/2}</param>
            </command_interface>
            <state_interface name="position">
                <param name="initial_value">${0}</param>
            </state_interface>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>

		<joint name="rleg1_ax3_joint">
            <command_interface name="position">
                <param name="min">${-pi}</param>
                <param name="max">${pi}</param>
            </command_interface>
            <state_interface name="position">
                <param name="initial_value">${0}</param>
            </state_interface>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>

		<joint name="rleg2_ax1_joint">
            <command_interface name="position">
                <param name="min">${-pi/2}</param>
                <param name="max">${pi/2}</param>
            </command_interface>
            <state_interface name="position">
                <param name="initial_value">${0}</param>
            </state_interface>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>

		<joint name="rleg2_ax2_joint">
            <command_interface name="position">
                <param name="min">${-pi/2}</param>
                <param name="max">${pi/2}</param>
            </command_interface>
            <state_interface name="position">
                <param name="initial_value">${0}</param>
            </state_interface>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>

		<joint name="rleg2_ax3_joint">
            <command_interface name="position">
                <param name="min">${-pi}</param>
                <param name="max">${pi}</param>
            </command_interface>
            <state_interface name="position">
                <param name="initial_value">${0}</param>
            </state_interface>
            <state_interface name="velocity"/>
            <state_interface name="effort"/>
        </joint>
	</ros2_control>

</robot>
